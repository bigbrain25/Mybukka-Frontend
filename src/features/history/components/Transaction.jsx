import React, { Fragment, useState, useEffect } from 'react';

import PropTypes from 'prop-types';

import Container from 'Components/container';
import { connect } from 'react-redux';
import ProfileHeaderTitle from '../../profile/common/ProfileHeaderTitle';
import trackingDisplayOpen from '../actionCreators/trackingDisplayOpen';
import Table from './Table';
import Card from './Card';
import Tracking from './Tracking';
import getOrderHistory from '../actionCreators/getOrderHistory';

const BgColor = ({ children }) => (
  <div className="bg-color history pt-2 pb-5">
    {children}
  </div>
);

const Transaction = ({
  userOrders,
  fetched,
  openTrackingDropdown,
  fetchedOrderHistory
}) => {
  const [refreshed, refresh] = useState(false);

  const extractPrice = dom => dom.cart.items
    .map(item => item.meal.price)
    .reduce((prev, curr) => prev + curr, 0);

  const extractQuantity = dom => dom.cart.items
    .map(item => item.quantity)
    .reduce((prev, curr) => prev + curr, 0);

  useEffect(() => {
    if (!refreshed && !fetched) {
      refresh(true);
      fetchedOrderHistory('/order?role=customer');
    }
  }, [refreshed, fetched]);

  return (
    <Fragment>
      <BgColor>
        <div className="profile-header-section mb-4">
          <ProfileHeaderTitle firstName="Order" lastName="history" />
        </div>
        {(userOrders && userOrders.length > 0) &&
        <Container classNames="relative d-bg-white-none">
          <div
            className="d-flex flex-column flex-xl-column
        flex-lg-column flex-md-column justify-content-between"
          >
            <Table
              extractPrice={extractPrice}
              extractQuantity={extractQuantity}
              data={userOrders}
              handleClick={openTrackingDropdown}
            />
            {userOrders.map(order => (
              <Card
                key={`order-transaction-card-${order._id}--${order.status}`}
                handleClick={() => openTrackingDropdown(order.status)}
                time={order.time}
                orderId={order._id} // eslint-disable-line
                mealTitle={order.cart.items[0].meal.title}
                price={extractPrice(order)}
                status={order.status}
                quantity={extractQuantity(order)}
              />
            ))}
          </div>
        </Container>
        }
      </BgColor>
      <Tracking />
    </Fragment>
  );
};

const mapStateToProps = ({
  getOrderHistoryReducer: { orderHistory: { userOrders }, status: { fetched, error } },
}) => ({
  fetched,
  error,
  userOrders
});

export default connect(
  mapStateToProps,
  {
    openTrackingDropdown: trackingDisplayOpen,
    fetchedOrderHistory: getOrderHistory
  }
)(Transaction);

Transaction.defaultProps = {
  userOrders: []
};

Transaction.propTypes = {
  fetched: PropTypes.bool.isRequired,
  fetchedOrderHistory: PropTypes.func.isRequired,
  openTrackingDropdown: PropTypes.func.isRequired,
  userOrders: PropTypes.arrayOf(
    PropTypes.objectOf(
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.objectOf(PropTypes.string)
      ])
    )
  )
};

BgColor.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
};
