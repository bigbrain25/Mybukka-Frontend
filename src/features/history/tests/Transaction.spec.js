import React from 'react';
import { Provider } from 'react-redux';
import Transaction from '../components/Transaction';

const initialState = {
  displayTrackingReducer: {
    show: true,
    status: 'pending',
  },
  getOrderHistoryReducer: { status: { fetched: true, error: false } },
};

const store = mockStore(initialState);


describe('Transaction component', () => {
  const props = {
    data: [{
      quantity: '2',
      title: 'salad and bbq',
      time: '27/3/2019 12:43 pm',
      orderId: '03d77',
      price: '3000',
      status: 'pending',
      courier: {
        name: 'curier name',
        img: 'https;//www.google.com/img',
      },
      pickup: {
        address: 'pickup address street',
        contactMobile: '03847858',
      },
      delivery: {
        address: 'delivery address street',
        name: 'mr delivery name',
        contactMobile: '94574783930',
      }
    }],
    handleClick: jest.fn(),
  };

  const { container, getByTestId } = render(
    <Provider store={store}>
      <Transaction {...props} />
    </Provider>
  );

  it('renders properly', () => {
    expect(container).toMatchSnapshot();
  });

  it('should triggers the handleclick function when clicked on table element', () => {
    const tableBtn = getByTestId('table-btn');
    fireEvent.click(tableBtn);
    expect(props.handleClick).toHaveBeenCalled();
  });

  it('should triggers the handleclick function when clicked on card', () => {
    const button = getByTestId('card-button');
    fireEvent.click(button);

    expect(props.handleClick).toBeCalled();
  });
});
