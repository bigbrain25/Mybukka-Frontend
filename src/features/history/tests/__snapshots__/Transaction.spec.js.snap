// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction component renders properly 1`] = `
<div>
  <div
    class="bg-color history pt-2 pb-5"
  >
    <div
      class="profile-header-section mb-4"
    >
      <div
        class="container profile-header"
      >
        <h1
          class="profile-header-title"
        >
          Order
           
          history
        </h1>
      </div>
    </div>
    <div
      class="container relative d-bg-white-none"
    >
      <div
        class="d-flex flex-column flex-xl-column
 flex-lg-column flex-md-column justify-content-between"
      >
        <div
          class="d-flex flex-column flex-xl-column flex-lg-column flex-md-column
 justify-content-between"
        >
          <div
            class="row flex custom-border d-xl-flex d-lg-flex d-md-none d-none"
          >
            <div
              class="col-2 pt-1 pb-1 custom-border bold"
            >
              Status
            </div>
            <div
              class="col-2 pt-1 pb-1 custom-border bold"
            >
              PickUp
            </div>
            <div
              class="col-2 pt-1 pb-1 custom-border bold"
            >
              DropOff
            </div>
            <div
              class="col-2 pt-1 pb-1 custom-border bold"
            >
              Order
            </div>
            <div
              class="col-2 pt-1 pb-1 custom-border bold"
            >
              Price
            </div>
            <div
              class="col-2 pt-1 pb-1 custom-border bold"
            >
              Courier
            </div>
          </div>
          <div
            aria-pressed="false"
            class="row flex flex-xl-row flex-lg-row d-xl-flex d-lg-flex d-md-none
 d-none cursor-pointer custom-border-px"
            data-testid="table-btn"
            role="button"
            tabindex="0"
          >
            <div
              class="col-xl-2 col-lg-2 col-md-10 pt-3 custom-border"
            >
              <div
                class="text-uppercase"
              >
                <h5
                  class="font-size-14 text-warning"
                >
                  pending
                </h5>
                <h5
                  class="light font-size-13 suggested-item-price"
                >
                  27/3/2019 12:43 pm
                </h5>
              </div>
            </div>
            <div
              class="col-xl-2 col-lg-2 col-md-10 pt-3 custom-border"
            >
              <div>
                <h5
                  class="text-uppercase font-size-14"
                >
                  salad and bbq
                </h5>
                <h5
                  class="light text-capitalize font-size-13"
                >
                  pickup address street
                </h5>
                <h5
                  class="light font-size-13"
                >
                  03847858
                </h5>
              </div>
            </div>
            <div
              class="col-xl-2 col-lg-2 col-md-10 pt-3 custom-border"
            >
              <div
                class="dropoff"
              >
                <h6
                  class="font-size-14 capitalize"
                >
                  mr delivery name
                </h6>
                <h6
                  class="light font-size-13"
                >
                  delivery address street
                </h6>
                <h6
                  class="light font-size-13"
                >
                  94574783930
                </h6>
              </div>
            </div>
            <div
              class="col-xl-2 col-lg-2 col-md-10 pt-3 custom-border"
            >
              <div
                class="history-order"
              >
                <h5
                  class="font-size-14"
                >
                  Order 
                  03d77
                </h5>
                <h5
                  class="light font-size-13"
                >
                  2
                   Items
                </h5>
              </div>
            </div>
            <div
              class="col-xl-2 col-lg-2 col-md-10 pt-3 custom-border"
            >
              <div
                class="history-price"
              >
                <h5
                  class="text-color font-size-14"
                >
                  ₦
                  3000
                </h5>
              </div>
            </div>
            <div
              class="col-xl-2 col-lg-2 col-md-10 pt-3 custom-border"
            >
              <div
                class="courier-history"
              >
                <img
                  alt="courier"
                  class="courier-img"
                  src="https;//www.google.com/img"
                />
                <h6
                  class="light"
                >
                  curier name
                </h6>
              </div>
            </div>
          </div>
        </div>
        <div
          aria-pressed="true"
          class="d-block d-md-block d-xl-none d-lg-none bg-white py-3 mb-3
 cursor-pointer"
          data-testid="card-button"
          role="button"
          tabindex="0"
        >
          <div
            class="container px-0 d-flex-column column mb-2 justify-content-center"
          >
            <div
              class="col-md-11 mx-auto "
            >
              <div
                class="row justify-content-between text-uppercase"
              >
                <h5
                  class="font-size-14"
                >
                  #
                  03d77
                </h5>
                <h5
                  class="font-size-14 suggested-item-price"
                >
                  27/3/2019 12:43 pm
                </h5>
              </div>
            </div>
            <hr
              class="border-line border-box col-12"
            />
            <div
              class="col-md-11 mx-auto pt-1 pb-1"
            >
              <div
                class="row justify-content-between"
              >
                <h5
                  class="text-uppercase font-size-14"
                >
                  2
                  X  
                  salad and bbq
                   ...
                </h5>
              </div>
            </div>
            <div
              class="col-md-11 mx-auto "
            >
              <div
                class="row justify-content-between text-capitalize"
              >
                <h5
                  class="font-size-14"
                >
                  Status: 
                  <span
                    class="text-warning"
                  >
                     
                    pending
                  </span>
                </h5>
                <h5
                  class="text-color font-size-14"
                >
                   ₦
                  3000
                  .00
                </h5>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="tracking-absolute tracking"
  >
    <div
      class="bg-color"
    >
      <div
        class="tracking-content content-bg-color"
      >
        <div
          class="
    position-fixed
    tracking-content-header"
        >
          <div
            class="row m-0 tracking-header"
          >
            <button
              class="tracking-cancel-btn"
              data-target=""
              data-testid="button"
              data-toggle=""
              id="Ztyq_yA84"
              type="button"
            >
              <svg
                fill="currentColor"
                height="24"
                viewBox="0 0 24 24"
                width="24"
              >
                <defs>
                  <path
                    d="M0 1.5L1.5 0 8 6.5 14.5 0 16 1.5 9.5 8l6.5 6.5-1.5 1.5L8 9.5 1.5 16 0 14.5 6.5 8z"
                    id="icon-close_svg__a"
                  />
                </defs>
                <use
                  transform="translate(4 4)"
                  xlink:href="#icon-close_svg__a"
                />
              </svg>
            </button>
            <h5>
              Logistics details
            </h5>
          </div>
          <hr
            class="text-muted m-0 border-box col-12"
          />
        </div>
        <div
          class="tracking-content-body"
        >
          <div
            class="timeline"
          >
            <div
              class="timeline-container timeline-left "
            >
              <div
                class="timeline-content"
              >
                <h5
                  class="font-size-14 suggested-item-price pt-2"
                >
                  5/3/16 10:55 AM
                </h5>
                <h5
                  class="font-size-14 suggested-item-price pb-3"
                >
                  order dispatched
                   
                  <i>
                    ✈...
                  </i>
                </h5>
              </div>
            </div>
          </div>
          <div
            class="timeline"
          >
            <div
              class="timeline-container timeline-left "
            >
              <div
                class="timeline-content"
              >
                <h5
                  class="font-size-14 suggested-item-price pt-2"
                >
                  5/3/16 10:55 AM
                </h5>
                <h5
                  class="font-size-14 suggested-item-price pb-3"
                >
                  Order received by the dispatch
                   
                  .
                </h5>
              </div>
            </div>
          </div>
          <div
            class="timeline"
          >
            <div
              class="timeline-container timeline-left timeline-active"
            >
              <div
                class="timeline-content"
              >
                <h5
                  class="font-size-14 suggested-item-price pt-2"
                >
                  5/3/16 10:55 AM
                </h5>
                <h5
                  class="font-size-14 suggested-item-price pb-3"
                >
                  waiting for pick up
                   
                  <i>
                    ✈...
                  </i>
                </h5>
              </div>
            </div>
          </div>
          <div
            class="timeline"
          >
            <div
              class="timeline-container timeline-left "
            >
              <div
                class="timeline-content"
              >
                <h5
                  class="font-size-14 suggested-item-price pt-2"
                >
                  5/3/16 10:55 AM
                </h5>
                <h5
                  class="font-size-14 suggested-item-price pb-3"
                >
                  Accepted by chef
                   
                  .
                </h5>
              </div>
            </div>
          </div>
          <div
            class="timeline"
          >
            <div
              class="timeline-container timeline-left "
            >
              <div
                class="timeline-content"
              >
                <h5
                  class="font-size-14 suggested-item-price pt-2"
                >
                  5/3/16 10:55 AM
                </h5>
                <h5
                  class="font-size-14 suggested-item-price pb-3"
                >
                  Your order was rejected by chef
                   
                  .
                </h5>
              </div>
            </div>
          </div>
          <div
            class="timeline"
          >
            <div
              class="timeline-container timeline-left "
            >
              <div
                class="timeline-content"
              >
                <h5
                  class="font-size-14 suggested-item-price pt-2"
                >
                  5/3/16 10:55 AM
                </h5>
                <h5
                  class="font-size-14 suggested-item-price pb-3"
                >
                  You cancelled this order
                   
                  .
                </h5>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="tracking-backdrop"
  />
</div>
`;
